File: spark/src/main/scala/geotrellis/spark/io/index/zcurve/ZSpaceTimeKeyIndex.scala
package geotrellis.spark.io.index.zcurve

import geotrellis.spark._
import geotrellis.spark.io.index.KeyIndex
import geotrellis.spark.io.index.zcurve._

import com.github.nscala_time.time.Imports._

object ZSpaceTimeKeyIndex {
  def byMilliseconds(keyBounds: KeyBounds[SpaceTimeKey], millis: Long): ZSpaceTimeKeyIndex =
    new ZSpaceTimeKeyIndex(keyBounds, millis)

  def bySecond(keyBounds: KeyBounds[SpaceTimeKey]): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L)

  def bySeconds(keyBounds: KeyBounds[SpaceTimeKey], seconds: Int): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * seconds)

  def byMinute(keyBounds: KeyBounds[SpaceTimeKey]): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60)

  def byMinutes(keyBounds: KeyBounds[SpaceTimeKey], minutes: Int): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60 * minutes)

  def byHour(keyBounds: KeyBounds[SpaceTimeKey]): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60 * 60)

  def byHours(keyBounds: KeyBounds[SpaceTimeKey], hours: Int): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60 * 60 * hours)

  def byDay(keyBounds: KeyBounds[SpaceTimeKey]): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60 * 60 * 24)

  def byDays(keyBounds: KeyBounds[SpaceTimeKey], days: Int): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60 * 60 * 24 * days)

  def byMonth(keyBounds: KeyBounds[SpaceTimeKey]): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60 * 60 * 24 * 30)

  def byMonths(keyBounds: KeyBounds[SpaceTimeKey], months: Int): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60 * 60 * 24 * 30 * months)

  def byYear(keyBounds: KeyBounds[SpaceTimeKey]): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60 * 60 * 24 * 365)

  def byYears(keyBounds: KeyBounds[SpaceTimeKey], years: Int): ZSpaceTimeKeyIndex =
    byMilliseconds(keyBounds, 1000L * 60 * 60 * 24 * 365 * years)
}

class ZSpaceTimeKeyIndex(val keyBounds: KeyBounds[SpaceTimeKey], val temporalResolution: Long) extends KeyIndex[SpaceTimeKey] {
  private def toZ(key: SpaceTimeKey): Z3 = Z3(key.col, key.row, (key.instant / temporalResolution).toInt)

  def toIndex(key: SpaceTimeKey): Long = toZ(key).z

  def indexRanges(keyRange: (SpaceTimeKey, SpaceTimeKey)): Seq[(Long, Long)] =
    Z3.zranges(toZ(keyRange._1), toZ(keyRange._2))
}

File: geotrellis-python/geotrellis/spark/io/index/zcurve/ZSpaceTimeKeyIndex.py
from geotrellis.spark.io.index.KeyIndex import KeyIndex
from geotrellis.spark.io.index.zcurve.Z3 import Z3
from geotrellis.spark.SpaceTimeKey import SpaceTimeKey

class ZSpaceTimeKeyIndex(KeyIndex[SpaceTimeKey]):
    def __init__(self, keyBounds, temporalResolution):
        self.keyBounds = keyBounds
        self.temporalResolution = temporalResolution

    def _toz(self, key):
        return Z3(key.col, key.row, int(key.instant / self.temporalResolution))

    def toIndex(self, key):
        return self._toz(key).z

    def indexRanges(self, keyRange):
        return Z3.zranges(self._toz(keyRange[0]), self._toz(keyRange[1]))

    @staticmethod
    def byMilliseconds(keyBounds, millis):
        return ZSpaceTimeKeyIndex(keyBounds, millis)

    @staticmethod
    def bySecond(keyBounds):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000)

    @staticmethod
    def bySeconds(keyBounds, seconds):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * seconds)

    @staticmethod
    def byMinute(keyBounds):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60)

    @staticmethod
    def byMinutes(keyBounds, minutes):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60 * minutes)

    @staticmethod
    def byHour(keyBounds):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60 * 60)

    @staticmethod
    def byHours(keyBounds, hours):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60 * 60 * hours)

    @staticmethod
    def byDay(keyBounds):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60 * 60 * 24)

    @staticmethod
    def byDays(keyBounds, days):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60 * 60 * 24 * days)

    @staticmethod
    def byMonth(keyBounds):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60 * 60 * 24 * 30)

    @staticmethod
    def byMonths(keyBounds, months):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60 * 60 * 24 * 30 * months)

    @staticmethod
    def byYear(keyBounds):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60 * 60 * 24 * 365)

    @staticmethod
    def byYears(keyBounds, years):
        return ZSpaceTimeKeyIndex.byMilliseconds(keyBounds, 1000 * 60 * 60 * 24 * 365 * years)
